File: Cpp/14.cpp
// Write a function to find the longest common prefix string amongst an array of
// strings.
//
// If there is no common prefix, return an empty string "".
//
// Example 1:
//
// Input: ["flower","flow","flight"]
// Output: "fl"
// Example 2:
//
// Input: ["dog","racecar","car"]
// Output: ""
// Explanation: There is no common prefix among the input strings.
// Note:
//
// All given inputs are in lowercase letters a-z.
#include<iostream>
#include<vector>
#include<string>

using namespace std;


class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        int n = strs.size();
        if (!n) {
            return "";
        }

        for (int k = 0; k < strs[0].size(); ++k) {
            for (int i = 1; i < n; ++i) {
                if (strs[i][k] != strs[0][k]) {
					cout<<strs[i][k]<<endl;
                    return strs[0].substr(0, k);
                }
            }
        }

        return strs[0];
    }
};

int main(){
	Solution s;
	vector<string> strs = {"hello", "he", "helloworld"};
	string res = s.longestCommonPrefix(strs);
	cout<<res<<endl;

	return 0;
}

File: Python/14.py
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        if not strs:
            return ""
        if len(strs) == 1:
            return strs[0]
        res = strs[0]
        for i in range(1, len(strs)):
            res = self.commonPrefix(res, strs[i])
        return res
            
    def commonPrefix(self, s1, s2):
        for i in range(len(s1)):
            if s1[:i+1] != s2[:i+1]:
                return s1[:i]
        return s1
            


class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        if not strs:
            return ""
        if len(strs) == 1:
            return strs[0]
        for i in range(len(strs[0])):
            for k in range(1, len(strs)):
                if i >= len(strs[k]) or strs[0][i] != strs[k][i]:
                    return strs[0][:i]
        return strs[0]
            
